### ISSUE TITLE
Json changes that used to compliment existing mutations now override them completely

### ISSUE THREAD

**Issue by dissociativity** [CleverRaven/Cataclysm-DDA#28508]
**Describe the bug**  
Json changes that used to compliment existing mutations now override them completely

**To Reproduce**  
Play with my maintained fork of PK rebalancing:
https://github.com/dissociativity/PKs_Rebalancing
Find that many mutations no longer work including and especially standard stat boost mutations, find that many mutation categories are utterly broken

**Expected behavior**  
The json changes in mutation overrides the mod add or change parts of existing vanilla mutations, instead they're totally replacing them with incomplete mutations/traits

**Versions and configuration(please complete the following information):**
 - OS: [Windows 8.1 Pro x74]
 - Game Version: Newest build as of 06/03/2019 (DD/MM/YYYY)
 - Graphics version: Tiles
 - Mods loaded: default plus PK re-balancing

**Comment by ifreund** [comment]
#27774 may be related

**Comment by kevingranade** [comment]
#27774 was definitely the cause, the override behaviour in the code was a bit too "subtle", and when #27774 "refactored" the code, that behaviour was lost.

**Comment by kevingranade** [comment]
On second look, #27774 seems to be doing the right thing with respect to handling overrides, generic_factory should be doing what we expect here.
After still more looking, what has happened is mutations switched from having an ad-hoc implicit override based on the id matching to a formal override as defined in generic_factory.

tl;dr AFAICT what you want to do is take every mutation in https://github.com/dissociativity/PKs_Rebalancing/blob/master/pk_mutation_override.json and replace:
```
"id": "foo",
```
with
```
"id": "foo",
"copy-from": "foo",
```

How it works in detail:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L179-L256

This may not be documented properly, but roughly how it works is:
If you have an "id" field, you are the one and only definition of your entity.
If you are in the core json, specifying the same "id" field twice is an error.
If you are in a mod and specify an already specified id (from either core or another mod), you silently override the existing entity.

If you specify a "copy-from" field, the contents of the specified entity (which must be of the same type, duh) are retrieved, and anything else you specify overwrites the existing values.
Important note, you do not have to make "copy-from" and "id" either be the same OR different.  If they are the same you replace the existing definition. If they are different you make a new item based on the existing item. Also if you specify the same "copy-from" and "id", the old entity is GONE, so for example if you want to have one entity based on "foo" and you want to have another entity REPLACE "foo", it's indeterminate which one gets evaluated first, so the only safe option is to have them both specify all the overrides they're supposed to share.

There's also an "abstract" which can be used instead of "id", this means that you only want the entity to exist for the purpose of inheritance by other entities, it will never have its own existence.  You can mix abstract and copy-from in the same ways as you can copy-from and id.  It should be an error to declare an abstract and never copy-from it into an id, but we might not catch that.

**Comment by dissociativity** [comment]
My only worry here is will copy-from make the mutation categories and thresholds work as before? being additive to the existing ones, or will they totally replace them so I'll need to add vanilla categories to them?

**Comment by ifreund** [comment]
Just tested and I can confirm that it does *not* append to the existing arrays as it did before.

**Comment by kevingranade** [comment]
For the time being you'll need to duplicate any field you override.

**Comment by kevingranade** [comment]
Upon further further investigation, it seems the integration with generic factory is incomplete, if the mutation load code specializes the appropriate generic reader classes, as outlined here:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L637-L669

It is possible that one of the existing helpers:
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L785
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L807
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L835
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/generic_factory.h#L855
Will do the right thing here.

**Comment by dobbry-vechur** [comment]
Cllosing as obsolete/expected behavior.
#77568 does not fully close #28520, but this issue is narrower: it expected duplicate mutation `id`s to partially merge with vanilla definitions. Current `generic_factory` behavior is that a duplicate `id` is a replacement unless `copy-from` is used.

Remaining gaps around documenting per-type/per-field inheritance support belong in #28520.
